x-pack/plugins/stack_alerts/server/alert_types/es_query/executor.ts

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { sha256 } from 'js-sha256';
+import { i18n } from '@kbn/i18n';
+import { CoreSetup, Logger } from 'kibana/server';
+import { addMessages, EsQueryAlertActionContext } from './action_context';
+import { ComparatorFns, getHumanReadableComparator } from '../lib';
+import { parseDuration } from '../../../../alerting/server';
+import { ExecutorOptions, OnlyEsQueryAlertParams, OnlySearchSourceAlertParams } from './types';
+import { ActionGroupId, ConditionMetAlertInstanceId } from './constants';
+import { fetchEsQuery } from './lib/fetch_es_query';
+import { EsQueryAlertParams } from './alert_type_params';
+import { fetchSearchSourceQuery } from './lib/fetch_search_source_query';
+import { Comparator } from '../../../common/comparator_types';
+
+export async function executor(
+  logger: Logger,
+  core: CoreSetup,
+  options: ExecutorOptions<EsQueryAlertParams>
+) {
+  const esQueryAlert = isEsQueryAlert(options);
+  const { alertId, name, services, params, state } = options;
+  const { alertFactory, scopedClusterClient, searchSourceClient } = services;
+  const currentTimestamp = new Date().toISOString();
+  const publicBaseUrl = core.http.basePath.publicBaseUrl ?? '';
+
+  const compareFn = ComparatorFns.get(params.thresholdComparator);
+  if (compareFn == null) {
+    throw new Error(getInvalidComparatorError(params.thresholdComparator));
+  }
+  let latestTimestamp: string | undefined = tryToParseAsDate(state.latestTimestamp);
+
+  // During each alert execution, we run the configured query, get a hit count
+  // (hits.total) and retrieve up to params.size hits. We
+  // evaluate the threshold condition using the value of hits.total. If the threshold
+  // condition is met, the hits are counted toward the query match and we update
+  // the alert state with the timestamp of the latest hit. In the next execution
+  // of the alert, the latestTimestamp will be used to gate the query in order to
+  // avoid counting a document multiple times.
+
+  const { numMatches, searchResult, dateStart, dateEnd } = esQueryAlert
+    ? await fetchEsQuery(alertId, name, params as OnlyEsQueryAlertParams, latestTimestamp, {
+        scopedClusterClient,
+        logger,
+      })
+    : await fetchSearchSourceQuery(
+        alertId,
+        params as OnlySearchSourceAlertParams,
+        latestTimestamp,
+        {
+          searchSourceClient,
+          logger,
+        }
+      );
+
+  // apply the alert condition
+  const conditionMet = compareFn(numMatches, params.threshold);
+
+  if (conditionMet) {
```

Comment:
nit: can it be an early return if `!conditionMet`?

Reply:
I would keep it for now and change that in follow up.